File: backend/app/models/TradePlan.py
from __future__ import annotations

from typing import Literal
from typing import TypedDict


class TradePlanInstrument(TypedDict):
    symbol: str
    market: str
    timeframe: str
    session: str


class TradePlanContext(TypedDict):
    bias: str
    trend: str
    phase: str
    narrative: str


class TradePlanConfirmation(TypedDict):
    trigger: str
    timeframe: str
    timestamp: str
    confluence: list[str]


class TradePlanEntry(TypedDict):
    orderType: Literal["MARKET", "LIMIT", "STOP"]
    price: float
    window: str


class TradePlanTargets(TypedDict):
    tp1: float
    tp2: float
    runner: str


class TradePlanRisk(TypedDict):
    stopLoss: float
    rr: str
    riskPercent: float


class TradePlanManagement(TypedDict):
    breakEven: str
    partialExit: str
    trailingRule: str


class TradePlanQuality(TypedDict):
    score: int
    grade: str


class TradePlanStatus(TypedDict):
    state: Literal["DRAFT", "READY", "ACTIVE", "INVALIDATED", "CLOSED"]
    updatedAt: str


class TradePlanChecklist(TypedDict):
    liquidityMapped: bool
    structureConfirmed: bool
    imbalanceAligned: bool
    riskValidated: bool
    sessionValidated: bool


class TradePlanInvalidation(TypedDict):
    price: float
    condition: str
    reason: str


class TradePlan(TypedDict):
    instrument: TradePlanInstrument
    context: TradePlanContext
    confirmation: TradePlanConfirmation
    entry: TradePlanEntry
    targets: TradePlanTargets
    risk: TradePlanRisk
    management: TradePlanManagement
    quality: TradePlanQuality
    status: TradePlanStatus
    checklist: TradePlanChecklist
    reasons: list[str]
    warnings: list[str]
    invalidation: TradePlanInvalidation

File: frontend/src/models/TradePlan.ts
export type TradePlanOrderType = "MARKET" | "LIMIT" | "STOP";

export type TradePlanState = "DRAFT" | "READY" | "ACTIVE" | "INVALIDATED" | "CLOSED";

export interface TradePlanInstrument {
  symbol: string;
  market: string;
  timeframe: string;
  session: string;
}

export interface TradePlanContext {
  bias: string;
  trend: string;
  phase: string;
  narrative: string;
}

export interface TradePlanConfirmation {
  trigger: string;
  timeframe: string;
  timestamp: string;
  confluence: string[];
}

export interface TradePlanEntry {
  orderType: TradePlanOrderType;
  price: number;
  window: string;
}

export interface TradePlanTargets {
  tp1: number;
  tp2: number;
  runner: string;
}

export interface TradePlanRisk {
  stopLoss: number;
  rr: string;
  riskPercent: number;
}

export interface TradePlanManagement {
  breakEven: string;
  partialExit: string;
  trailingRule: string;
}

export interface TradePlanQuality {
  score: number;
  grade: string;
}

export interface TradePlanStatus {
  state: TradePlanState;
  updatedAt: string;
}

export interface TradePlanChecklist {
  liquidityMapped: boolean;
  structureConfirmed: boolean;
  imbalanceAligned: boolean;
  riskValidated: boolean;
  sessionValidated: boolean;
}

export interface TradePlanInvalidation {
  price: number;
  condition: string;
  reason: string;
}

export interface TradePlan {
  instrument: TradePlanInstrument;
  context: TradePlanContext;
  confirmation: TradePlanConfirmation;
  entry: TradePlanEntry;
  targets: TradePlanTargets;
  risk: TradePlanRisk;
  management: TradePlanManagement;
  quality: TradePlanQuality;
  status: TradePlanStatus;
  checklist: TradePlanChecklist;
  reasons: string[];
  warnings: string[];
  invalidation: TradePlanInvalidation;
}
